test/object-store/results.cpp

```diff
@@ -3525,3 +3525,106 @@ TEST_CASE("results: query helpers", "[include]") {
+    _impl::RealmCoordinator::assert_no_open_realms();
+
+    InMemoryTestFile config;
+    config.cache = false;
+    config.automatic_change_notifications = false;
+
+    auto r = Realm::get_shared_realm(config);
+    r->update_schema({
+        {"no_pk",
+         {
+             {"id", PropertyType::Int},
+             {"value", PropertyType::Int},
+         }},
+        {"int_pk",
+         {
+             {"id", PropertyType::Int, Property::IsPrimary{true}},
+             {"value", PropertyType::Int},
+         }},
+        {"string_pk",
+         {
+             {"id", PropertyType::String, Property::IsPrimary{true}},
+             {"value", PropertyType::Int},
+         }},
+    });
+
+    auto add_callback = [](Results& results, int& calls, CollectionChangeSet& changes) {
+        return results.add_notification_callback([&](CollectionChangeSet c, std::exception_ptr err) {
+            REQUIRE_FALSE(err);
+            ++calls;
+            changes = std::move(c);
+        });
+    };
+
+    auto coordinator = _impl::RealmCoordinator::get_existing_coordinator(config.path);
+    auto table1 = r->read_group().get_table("class_no_pk");
+    auto table2 = r->read_group().get_table("class_int_pk");
+    auto table3 = r->read_group().get_table("class_string_pk");
+
+    TestContext d(r);
+    auto create = [&](StringData type, util::Any&& value) {
+        return Object::create(d, r, *r->schema().find(type), value);
+    };
+
+    r->begin_transaction();
+    auto k1 = create("no_pk", AnyDict{{"id", INT64_C(123)}, {"value", INT64_C(100)}}).obj().get_key();
+    auto k2 = create("int_pk", AnyDict{{"id", INT64_C(456)}, {"value", INT64_C(100)}}).obj().get_key();
+    auto k3 = create("string_pk", AnyDict{{"id", std::string("hello")}, {"value", INT64_C(100)}}).obj().get_key();
+    r->commit_transaction();
+
+    Results results1(r, table1->where());
+    int calls1 = 0;
+    CollectionChangeSet changes1;
+    auto token1 = add_callback(results1, calls1, changes1);
+
+    Results results2(r, table2->where());
+    int calls2 = 0;
+    CollectionChangeSet changes2;
+    auto token2 = add_callback(results2, calls2, changes2);
+
+    Results results3(r, table3->where());
+    int calls3 = 0;
+    CollectionChangeSet changes3;
+    auto token3 = add_callback(results3, calls3, changes3);
+
+    advance_and_notify(*r);
+    REQUIRE(calls1 == 1);
+    REQUIRE(calls2 == 1);
+    REQUIRE(calls3 == 1);
+
+    SECTION("objects removed") {
+        r->begin_transaction();
+        r->read_group().get_table("class_no_pk")->remove_object(k1);
+        r->read_group().get_table("class_int_pk")->remove_object(k2);
+        r->read_group().get_table("class_string_pk")->remove_object(k3);
+        create("no_pk", AnyDict{{"id", INT64_C(123)}, {"value", INT64_C(200)}});
+        create("int_pk", AnyDict{{"id", INT64_C(456)}, {"value", INT64_C(200)}});
+        create("string_pk", AnyDict{{"id", std::string("hello")}, {"value", INT64_C(200)}});
```

Comment:
What is the expected notification changes for this event? Can you add a check that this is being observed as `CollectionChangeSet::modifications` and not a deletion and insertion? at least I think that is the expected behaviour, except for maybe the `no_pk` case where we don't guarantee reuse of auto generated pks?

Reply:
This should be reported as a delete and insert. All of these tests need to be validating the changeset and not just that a notification was sent.

Reply:
The test is now enhanced with those tests.